### PR TITLE
#74: Added refresh request.

### LINKED ISSUE #74: Refresh Request
# Situation

At the moment the refresh request is not supported.

# Acceptance Criteria

1. Virtual Schema can be refreshed with [`ALTER VIRTUAL SCHEMA ... REFRESH`](https://docs.exasol.com/sql/alter_schema.htm#microcontent2)

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #74.

### CODE REVIEW

File: .classpath [hunk 1]
```diff
@@ -9,6 +9,7 @@
 	</classpathentry>
```
Comment (kaklakariada) [inline .classpath:9]: .classpath should be removed and ignored. Eclipse will generate it from the pom

File: doc/changes/changes_0.5.0.md [hunk 2]
```diff
@@ -1,11 +1,13 @@
 # row-level-security-lua 0.5.0, released 2021-09-01
 
-Code name: Support for Default SRID
+Code name: Schema refresh, exclude capabilities and default SRID
 
 ## Summary
 
 Version 0.5.0 of `row-level-security` adds support for `GEOMETRY` columns without explicit SRID.
 
+Lua RLS now supports `ALTER VIRTUAL SCHEMA ... REFRESH` to update the metadata of an existing virtual schema.
```
Comment (kaklakariada) [inline doc/changes/changes_0.5.0.md:9]: Is the `RLS` abbreviation well known outside Exasol?

File: src/main/lua/exasolrls/rls_adapter.lua [hunk 3]
```diff
@@ -49,7 +49,23 @@ function M.drop_virtual_schema()
...
 end
 
-function M.refresh()
+---
+-- Refresh the metadata of the Virtual Schema.
+-- <p>
+-- Re-reads the structure and data types of the schema protected by RLS.
+-- </p>
+--
+-- @param exa_metadata Exasol metadata (not used)
```
Comment (kaklakariada) [inline src/main/lua/exasolrls/rls_adapter.lua:58]:
```suggestion
-- @param _ Exasol metadata (not used)
```
the parameter name is `_` (only mentioned once)

File: src/test/lua/test_text.lua [hunk 4]
```diff
@@ -8,12 +8,12 @@ function test_text.test_trim()
 end
 
 function test_text.test_split()
-    tests ={
+    local tests ={
```
Comment (kaklakariada) [inline src/test/lua/test_text.lua:11]:
```suggestion
    local tests = {
```

File: src/test/lua/test_text.lua [hunk 5]
```diff
@@ -8,12 +8,12 @@ function test_text.test_trim()
...
         {
-            input = "foo,bar,baz", expected = {"foo", "bar", "baz"},
-            input = "foo,,bar,baz", expected = {"foo", "bar", "baz"},
-            input = ",1,2,3", expected = {"1", "2", "3"},
-            input = " Hello , world    , ! ", expected = {"Hello", "world", "!"}
+            {input = "foo,bar,baz", expected = {"foo", "bar", "baz"}},
+            {input = "foo,,bar,baz", expected = {"foo", "bar", "baz"}},
+            {input = ",1,2,3", expected = {"1", "2", "3"}},
+            {input = " Hello , world    , ! ", expected = {"Hello", "world", "!"}}
```
Comment (kaklakariada) [inline src/test/lua/test_text.lua:16]: Please add test with `nil` input

File: src/test/java/com/exasol/MetadataReadingIT.java [hunk 6]
```diff
@@ -21,26 +24,22 @@
...
         createGroupToUserMappingTable(sourceSchema) //
-                .insert(groupName, userName);
+                .insert(groupName, "USER_FOR_LATE_REGISTERED_TABLE");
         sourceSchema.createTable("T", "C1", "BOOLEAN", ROW_GROUP_COLUMN, IDENTIFIER_TYPE) //
                 .insert(true, groupName) //
                 .insert(false, "GROUP_THE_USER_DOES_NOT_HAVE");
         final VirtualSchema virtualSchema = createVirtualSchema(sourceSchema);
-        final User user = createUserWithVirtualSchemaAccess(userName, virtualSchema);
+        final User user = createUserWithVirtualSchemaAccess("USER_FOR_LATE_REGISTERED_TABLE", virtualSchema);
         assertRlsQueryWithUser("SELECT C1 FROM " + virtualSchema.getName() + ".T", user, table().row(true).matches());
```
Comment (kaklakariada) [inline src/test/java/com/exasol/MetadataReadingIT.java:36]: Does this verify, that the query returns a single row?

File: src/test/java/com/exasol/MetadataReadingIT.java [hunk 7]
```diff
@@ -61,27 +60,85 @@ void testDetermineColumnTypes() {
...
+
+    private void assertVirtualTableStructure(final Table table, final User user,
+            final Matcher<ResultSet> tableMatcher) {
+        assertRlsQueryWithUser("DESCRIBE " + getVirtualSchemaName(table.getParent().getName()) + "." + table.getName(),
+                user, //
+                tableMatcher);
+    }
+
+    private Matcher<ResultSet> expectRows(final String... strings) {
+        assert ((strings.length % 2) == 0);
```
Comment (kaklakariada) [inline src/test/java/com/exasol/MetadataReadingIT.java:92]: better use a junit assertion with a meaningful error message in case `-ea` is not used.

File: src/test/java/com/exasol/MetadataReadingIT.java [hunk 8]
```diff
@@ -61,27 +60,85 @@ void testDetermineColumnTypes() {
...
+        refreshVirtualSchema(virtualSchema);
+        assertVirtualTableStructure(modifiedTable, user, expectRows("VU", "VARCHAR(40) UTF8", "DO", "DOUBLE"));
+    }
+
+    private void refreshVirtualSchema(final VirtualSchema virtualSchema) {
+        final String sql = "ALTER VIRTUAL SCHEMA " + virtualSchema.getFullyQualifiedName() + " REFRESH";
+        try {
+            execute(sql);
+        } catch (final SQLException exception) {
+            throw new AssertionError("Unable to refresh Virtual Schema.", exception);
```
Comment (kaklakariada) [inline src/test/java/com/exasol/MetadataReadingIT.java:119]:
```suggestion
            throw new AssertionError("Unable to refresh Virtual Schema using query '" + sql + "'"., exception);
```

**Review by kaklakariada — CHANGES_REQUESTED** [review]

**Comment by kaklakariada** [comment]
There are some dependency updates available. Also upgrade Exasol to 7.1.1

File: .classpath (hunk 1 above)
Comment (redcatbear) [reply .classpath:9]: Not with hybrid projects unfortunately. And this is a Lua / Java hybrid.

File: doc/changes/changes_0.5.0.md (hunk 2 above)
Comment (redcatbear) [reply doc/changes/changes_0.5.0.md:9]: Yes, but I wrote it out anyway. Just in case.

File: src/main/lua/exasolrls/rls_adapter.lua (hunk 3 above)
Comment (redcatbear) [reply src/main/lua/exasolrls/rls_adapter.lua:58]: Fixed, good find!

**Review by kaklakariada — APPROVED** [review]
